### PR TITLE
fix(android): properly handle selection 🎢

### LINKED ISSUE #6177: bug: iOS and Android Text Select weird behavior
**Describe the bug**
Selecting text in iOS and Android with Keyman keyboard is resulting in weird results
in Android, selecting text will copy and paste text
in iOS, text is deleted

**To Reproduce**
1. with Keyman keyboard, highlight any text in notes app or whatsapp
2. Text will be copied and pasted in Android, Text will be deleted in iOS

**Expected behavior**
Show context menu

**Keyman for iPhone:**
 - Device: iPhone 13 Pro Max
 - OS: iOS 15.2.1
 - Keyman version 15.0.186-alpha
 - Target application: Notes, Whatsapp

**Keyboard**
 - Keyboard name: EuroSil

### LINKED ISSUE #6261: bug:
**Describe the bug**

**To Reproduce**

**Expected behavior**

**Screenshots**

https://user-images.githubusercontent.com/4498365/153778577-34a14255-09eb-4946-b672-8cf67ba998b5.mp4
---

**Keyman for Windows/macOS/Linux:**
 - OS: 
 - Keyman version: 
 - Target Application: 

**Keyman for iPhone/iPad/Android:**
 - Device: 
 - OS: 
 - Keyman version 
 - Target application: 

**KeymanWeb:**
 - OS: 
 - Browser: 
 - KeymanWeb Version: 
 - URL: 

**Keyman Developer:**
 - OS: 
 - Keyman Developer version: 

---

**Keyboard**
 - Keyboard name: 
 - Keyboard version: 
 - Language name: 

**Additional context**

### PR DESCRIPTION
**Pull request by mcdurdin**
Part of #5853.

Fixes #6261.
Fixes #6177 (Android).

Building on the fixes in KeymanWeb in #6272, we do similar patches for Android.

I think it is likely that the iOS tests will fail -- in which case, I will work through similar patches for iOS.

# User Testing

(Note: these tests are very good candidates for general acceptance tests!)

Where requested, these tests should be run using the attached [web_context_tests keyboard](https://drive.google.com/file/d/1C1I39AxR9egfuQKXRbwwyK4T9YBw6Ji-/view?usp=sharing).

* **GROUP_ANDROID_INAPP:** Test on Keyman for Android, in-app
* **GROUP_ANDROID_SYSTEM:** Test on Keyman for Android, system keyboard

**TEST_BASELINE:** Using a keyboard with some complex behaviours, such as Khmer Angkor, verify that input sequences are correct. For example, type <kbd>x</kbd><kbd>E</kbd><kbd>j</kbd><kbd>m</kbd><kbd>r</kbd> (desktop: <kbd>ខ</kbd><kbd>ែ</kbd><kbd>្</kbd><kbd>ម</kbd><kbd>រ</kbd> touch: <kbd>ខ</kbd><kbd>ែ (longpress េ)</kbd><kbd>្ម (longpress ម)</kbd><kbd>រ</kbd>) and verify that the emitted output is actually `ខ្មែរ` (you may need to use a tool to view the Unicode values, which should be as in the image below).

image

**TEST_SELECTION:** Using the web_context_tests keyboard, type <kbd>a</kbd><kbd>b</kbd><kbd>c</kbd><kbd>x</kbd>. Select `bc` and type <kbd>q</kbd>. The output after the final key should be `aqx`. (This is testing selection replacement with no matching keyboard rule.)

**TEST_SELECTION_2:** Using the web_context_tests keyboard, type <kbd>a</kbd><kbd>b</kbd><kbd>c</kbd><kbd>x</kbd>. Select `bc` and type <kbd>p</kbd>. The output after the final key should be `aqx`. (This is testing selection replacement with a matching keyboard rule.)

**TEST_CONTEXT_BASE:** Using the web_context_tests keyboard, type <kbd>a</kbd><kbd>b</kbd><kbd>c</kbd><kbd>d</kbd>. The output after the final key should be `!`.

**TEST_CONTEXT_SELECTION:** Using the attached web_context_tests keyboard, type <kbd>a</kbd><kbd>b</kbd><kbd>c</kbd><kbd>x</kbd>. Select the `x` character, and type <kbd>d</kbd>. The output after the final key should be `abcd`.

**TEST_CONTEXT_SELECTION_2:** Using the attached web_context_tests keyboard, type <kbd>a</kbd><kbd>b</kbd><kbd>c</kbd><kbd>x</kbd>. Select the `x` character, and delete it with <kbd>Backspace</kbd>. Type <kbd>d</kbd>. The output after the final key should be `!`.

**TEST_CONTEXT_SELECTION_3:** Using the attached web_context_tests keyboard, type <kbd>a</kbd><kbd>b</kbd><kbd>c</kbd><kbd>x</kbd>. Select the `x` character, and type <kbd>y</kbd>. Press <kbd>Backspace</kbd>, and type <kbd>d</kbd>. The output after the final key should be `!`.

**TEST_CONTEXT_SELECTION_4:** Using the attached web_context_tests keyboard, type <kbd>x</kbd><kbd>a</kbd><kbd>b</kbd><kbd>c</kbd><kbd>x</kbd>. Select the `abc` characters, and type <kbd>d</kbd>. The output after the final key should be `xdx`.

## SMP Tests

**TEST_SELECTION_SMP:** Using the web_context_tests keyboard, type <kbd>f</kbd><kbd>g</kbd><kbd>h</kbd><kbd>j</kbd>. Select `𐌁𐌂` and type <kbd>o</kbd>. The output after the final key should be `𐌀𐍈𐌈`.

**TEST_CONTEXT_BASE_SMP:** Using the web_context_tests keyboard, type <kbd>f</kbd><kbd>g</kbd><kbd>h</kbd><kbd>i</kbd>. The output after the final key should be `𐌙`.

**TEST_CONTEXT_SELECTION_SMP:** Using the attached web_context_tests keyboard, type <kbd>f</kbd><kbd>g</kbd><kbd>h</kbd><kbd>j</kbd>. Select the `𐌈` character, and type <kbd>i</kbd>. The output after the final key should be `𐌀𐌁𐌂i`.

**TEST_CONTEXT_SELECTION_SMP_2:** Using the attached web_context_tests keyboard, type <kbd>f</kbd><kbd>g</kbd><kbd>h</kbd><kbd>j</kbd>. Select the `𐌈` character, and delete it with <kbd>Backspace</kbd>. Type <kbd>i</kbd>. The output after the final key should be `𐌙`.

**TEST_CONTEXT_SELECTION_SMP_3:** Using the attached web_context_tests keyboard, type <kbd>f</kbd><kbd>g</kbd><kbd>h</kbd><kbd>j</kbd>. Select the `𐌈` character, and type <kbd>k</kbd>. Press <kbd>Backspace</kbd>, and type <kbd>i</kbd>. The output after the final key should be `𐌙`.

**TEST_CONTEXT_SELECTION_SMP_4:** Using the attached web_context_tests keyboard, type <kbd>k</kbd><kbd>f</kbd><kbd>g</kbd><kbd>h</kbd><kbd>k</kbd>. Select the `𐌀𐌁𐌂` characters, and type <kbd>i</kbd>. The output after the final key should be `𐌰i𐌰`.

### CODE REVIEW

File: android/KMEA/app/src/main/assets/android-host.js [hunk 1]
```diff
@@ -1,3 +1,10 @@
+let _debug = 0;
```
Comment (darcywong00) [inline android/KMEA/app/src/main/assets/android-host.js:1]:
```suggestion
var _debug = 0;
```

Unfortunately, the `let` syntax isn't supported by Chrome on our older Android 5.0 devices (SDK 21).
Also for l. 227

**Comment by darcywong00** [comment]
Hmm, from my cursory testing of ANDROID_INAPP: (API 29 emulator with default sil_euro_latin keyboard)

1. enter some text ("the quick brown fox")
2. select the entire text
3. move the start cursor to the start of "quick" (selection to the end still maintained
4. type a key (e.g. <kbd>h</kbd>
5. the selected text remains

**Comment by mcdurdin** [comment]
> Hmm, from my cursory testing of ANDROID_INAPP: (API 29 emulator with default sil_euro_latin keyboard)

Yeah. I don't know how I got that wrong... Have a go now with the latest build, just landing now.

**Comment by darcywong00** [comment]
ok - I checked out how SMP is handled for in-app keyboard with basic_kbdosm version 1.1
It seems when the selection is deleted, it's deleting more on the left...

1. Install basic_kbdosm version 1.1
2. Type the keys <kbd>qwerty</kbd> <kbd>enter</kbd> <kbd>qwerty</kbd>
3. Select as below
osmanya setup

4. Type key <kbd>h</kbd>

5. Observe more than the selected text is deleted
osmanya result




and then I hit <kbd>h</kbd>, the deletion goes into the previous line and the inserted text added there.

**Comment by mcdurdin** [comment]
> But then I checked out how SMP is handled for in-app keyboard with basic_kbdosm version 1.1

Good catch. I've updated the web_context_tests keyboard to add SMP tests. Am about to update the user test instructions to add those!

**Comment by rc-swag** [comment]
GROUP_ANDROID_SYSTEM
TEST_CONTEXT_BASE_SMP: FAIL 
TEST_CONTEXT_SELECTION_SMP_2: FAIL 
TEST_CONTEXT_SELECTION_SMP_3: FAIL
Result: ?𐌙�
When you use backspace you can not remove the � it just replace the 𐌙 with a � the only way to remove the �'s is by selecting the text pressing <kbd>backspace</kbd>

test_failsmp

**Comment by rc-swag** [comment]
On the TEST_CONTEXT_SELECTION_SMP_3 in GROUP_ANDROID_INAPP the test passes. However I think midway through the steps I get an unexpected output.
Using the attached web_context_tests keyboard, type <kbd>f</kbd><kbd>g</kbd><kbd>h</kbd><kbd>j</kbd>. Select the `𐌈` character, and type <kbd>k</kbd>
I thought it should be xxx but it is xxx

**Comment by MakaraSok** [comment]
All tests done on Galaxy S9+ device in Android 10 (One UI 2.5).

* GROUP_ANDROID_INAPP: Test on Keyman for Android, in-app
  - **TEST_BASELINE (PASSED)** The output is indeed the expected `ខ្មែរ`.
  - **TEST_SELECTION (PASSED)** The output is indeed the expected `aqx`.
  aqx<br/>
  - **TEST_CONTEXT_BASE (PASSED)** The output is indeed the expected one as indicated in the instructions.
  - **TEST_CONTEXT_SELECTION (PASSED)** The output is indeed the expected one as indicated in the instructions.
  - **TEST_CONTEXT_SELECTION_2 (PASSED)** The output is indeed the expected one as indicated in the instructions.
  - **TEST_CONTEXT_SELECTION_3 (PASSED)** The output is indeed the expected one as indicated in the instructions.
  - **TEST_CONTEXT_SELECTION_4 (PASSED)** The output is indeed the expected one as indicated in the instructions.
  - **TEST_SELECTION_SMP (BLOCKED)** The instructions may be erroneous. 
  - **TEST_CONTEXT_BASE_SMP (BLOCKED)** The instructions may be erroneous. 
  - **TEST_CONTEXT_SELECTION_SMP (BLOCKED)** The instructions may be erroneous. 
  - **TEST_CONTEXT_SELECTION_SMP_2 (BLOCKED)** The instructions may be erroneous. 
  - **TEST_CONTEXT_SELECTION_SMP_3 (BLOCKED)** The instructions may be erroneous. 
  - **TEST_CONTEXT_SELECTION_SMP_4 (BLOCKED)** The instructions may be erroneous. 
  erroneous instructions<br/>

* GROUP_ANDROID_SYSTEM: Test on Keyman for Android, system keyboard
  - **TEST_BASELINE (PASSED)** The output is indeed the expected `ខ្មែរ`.
  - **TEST_SELECTION (PASSED)** The output is indeed the expected `aqx`.
  image<br/>
  - **TEST_CONTEXT_BASE (PASSED)** The output is indeed the expected one as indicated in the instructions.
  - **TEST_CONTEXT_SELECTION (PASSED)** The output is indeed the expected one as indicated in the instructions.
  - **TEST_CONTEXT_SELECTION_2 (PASSED)** The output is indeed the expected one as indicated in the instructions.
  - **TEST_CONTEXT_SELECTION_3 (PASSED)** The output is indeed the expected one as indicated in the instructions.
  - **TEST_CONTEXT_SELECTION_4 (PASSED)** The output is indeed the expected one as indicated in the instructions.
  - **TEST_SELECTION_SMP (OPEN)**
  - **TEST_CONTEXT_SELECTION_SMP (BLOCKED)** The instructions may be erroneous. 
  - **TEST_CONTEXT_SELECTION_SMP_4 (BLOCKED)** The instructions may be erroneous. 
    erroneous instructions<br/>

**Comment by mcdurdin** [comment]
> * **TEST_SELECTION_SMP (BLOCKED)** The instructions may be erroneous.

Can you give me some indication of what you are having trouble with in the instructions? Note that if you have downloaded the .kmp earlier, you should download again to get the latest version of it (it changed with the most recent commits).

**Comment by MakaraSok** [comment]
All tests done on Galaxy S9+ device in Android 10 (One UI 2.5).

* GROUP_ANDROID_INAPP: Test on Keyman for Android, in-app
  - **TEST_SELECTION_SMP (FAILED)** The output is `𐍈𐌈` instead of the expected `𐌀𐍈𐌈`. Note though that `𐌂` in the instructions is the other way around.
  reversed c?<br/>
  - **TEST_CONTEXT_BASE_SMP (PASSED)** The output is `𐌙` as expected.
  - **TEST_CONTEXT_SELECTION_SMP (PASSED)** The output is `𐌀𐌁𐌂i` as expected. Note though that `𐌂` in the instructions is the other way around.
  reversed c?<br/>
  - **TEST_CONTEXT_SELECTION_SMP_2 (PASSED)** The output is `𐌙` as expected.
  - **TEST_CONTEXT_SELECTION_SMP_3 (PASSED)** The output is `𐌙` as expected.
  - **TEST_CONTEXT_SELECTION_SMP_4 (PASSED)** The output is `𐌰i𐌰` as expected.

* GROUP_ANDROID_SYSTEM: Test on Keyman for Android, system keyboard
  - **TEST_SELECTION_SMP (PASSED)** The output is shown as indicated.
  - **TEST_CONTEXT_BASE_SMP (FAILED)** The output is not shown as indicated. There are additional characters around the expected character output.
  additional characters<br/>
  - **TEST_CONTEXT_SELECTION_SMP (PASSED)** The output is `𐌀𐌁𐌂i` as expected. Note though that `𐌂` in the instructions is the other way around.
  image<br/>
  - **TEST_CONTEXT_SELECTION_SMP_2 (FAILED)** The output is not shown as indicated. There are additional characters around the expected character output.
  additional characters<br/>
  - **TEST_CONTEXT_SELECTION_SMP_3 (FAILED)** The output is not shown as indicated. There are additional characters around the expected character output.
  additional characters<br/>
  - **TEST_CONTEXT_SELECTION_SMP_4 (PASSED)** The output is shown as indicated.
  image<br/>

**Comment by mcdurdin** [comment]
I've fixed two separate issues in the text selection code and teseted here. I think we need to retest all the scenarios now. Sorry!

@keymanapp-test-bot retest

**Comment by darcywong00** [comment]
## SUITE_INAPP: Test inapp group

* GROUP_ANDROID_INAPP: Test on Keyman for Android, in-app

  - **TEST_BASELINE: PASSED** Khmer Angkor keyboard outputs `ខ្មែរ`
  - **TEST_SELECTION: PASSED** Result is `aqx`
  - **TEST_SELECTION_2: PASSED** Result is `aqx`
  - **TEST_CONTEXT_BASE: PASSED** Result is `!`
  - **TEST_CONTEXT_SELECTION: PASSED** Result is `abcd`
  - **TEST_CONTEXT_SELECTION_2: PASSED** Result is `!`
  - **TEST_CONTEXT_SELECTION_3: PASSED** Result is `!`
  - **TEST_CONTEXT_SELECTION_4: PASSED** Result is `xdx`
  - **TEST_SELECTION_SMP: PASSED** Result is `𐌀𐍈𐌈`
  - **TEST_CONTEXT_BASE_SMP: PASSED** Result is `𐌙`
  - **TEST_CONTEXT_SELECTION_SMP: PASSED** Result is `𐌀𐌁𐌂i`
  - **TEST_CONTEXT_SELECTION_SMP_2: PASSED** Result is `𐌙`
  - **TEST_CONTEXT_SELECTION_SMP_3: PASSED** Result is `𐌙`
  - **TEST_CONTEXT_SELECTION_SMP_4: PASSED** Result is `𐌰i𐌰`

File: android/KMEA/app/src/main/assets/android-host.js [hunk 2]
```diff
@@ -1,3 +1,10 @@
+var _debug = 1;
```
Comment (darcywong00) [inline android/KMEA/app/src/main/assets/android-host.js:1]: Do we want to check this in as `0`?

Comment (mcdurdin) [inline android/KMEA/app/src/main/assets/android-host.js:1]:
```suggestion
var _debug = 0;
```

Let's turn off debug logging before merge 😁

Comment (mcdurdin) [reply android/KMEA/app/src/main/assets/android-host.js:1]: *snap* -- see my comment at exact same time!

File: android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMManager.java [hunk 3]
```diff
@@ -2925,7 +2925,17 @@ private static void performLeftDeletions(InputConnection ic, int dn) {
...
+      // Count the number of characters which are surrogate pairs
+      int index = lastIndex, dnx = dn, numPairs = 0;
+      while(index > 0 && dnx > 0) {
+        if(Character.isLowSurrogate(charsBackup.charAt(index)) &&
+            Character.isHighSurrogate(charsBackup.charAt(index-1))) {
+          numPairs++;
+          index--;
+        }
+        index--;
+        dnx--;
```
Comment (darcywong00) [inline android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMManager.java:2937]: I read the commit message that countSurrogatePairs() was incorrect for system keyboard, but that it was used elsewhere correctly.

Should we document the limitation in CharSequenceUtil.countSurrogatePairs()? (I'm also still trying to understand the difference)

**Review by darcywong00 — APPROVED** [review]
lgtm

File: android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMManager.java (hunk 3 above)
Comment (mcdurdin) [reply android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMManager.java:2937]: `CharSequenceUtil.countSurrogatePairs()` counts the number of surrogate pairs in a string of `n` code units (i.e. 16 bit chars) length. This is counting the number of surrogate pairs in a string of `n` code points (which may be either single 16 bit chars or surrogate pairs).

Comment (rc-swag) [reply android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMManager.java:2937]: I still don't get why `CharSequenceUtil.countSurrogatePairs()`  is considered to be working fine? It didn't work which is why it was replaced with a inline solution so how can we have confidence it is working?

**Comment by rc-swag** [comment]
I should also add that the user tests work for me now.

File: android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMManager.java (hunk 3 above)
Comment (mcdurdin) [reply android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMManager.java:2937]: Continuing the `countSurrogatePairs()` discussion in #6301. Needs review but I want to try and get this selection stuff sorted.

**Comment by rc-swag** [comment]
lgtm

**Review by rc-swag — APPROVED** [review]
lgtm

**Pull request merged**